samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+		if (err < 0) {
+			return err;
+		}
+		err = at_params_short_get(&at_param_list, 5, &retain);
+		if (err < 0) {
+			return err;
+		}
+		if (datatype == DATATYPE_HEXADECIMAL) {
+			u8_t *data_hex;
+
+			data_hex = k_malloc(msg_sz / 2);
+			if (!data_hex) {
+				return -ENOMEM;
+			}
+			err = slm_util_atoh(msg, msg_sz,
+						data_hex, msg_sz / 2);
+			if (err > 0) {
+				err = do_mqtt_publish(qos, retain,
+							topic, topic_sz,
+							data_hex, err);
+			}
+			k_free(data_hex);
+		} else {
+			err = do_mqtt_publish(qos, retain,
+						topic, topic_sz,
+						msg, msg_sz);
+			if (err < 0) {
+				return err;
+			}
+		}
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf, "#XMQTTPUB: <topic>, (0, 1), <msg>,"
+					" (0, 1, 2), (0, 1)\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XMQTTSUB commands
+ *  AT#XMQTTSUB=<topic>,<op>[,<qos>]
+ *  AT#XMQTTSUB? READ command not supported
+ *  AT#XMQTTPUB=?
+ */
+static int handle_at_mqtt_subscribe(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	u16_t op, qos;
+	char topic[MQTT_MAX_TOPIC_LEN];
+	int topic_sz = MQTT_MAX_TOPIC_LEN;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) == 4) {
+			err = at_params_string_get(&at_param_list, 1,
+							topic, &topic_sz);
+			if (err < 0) {
+				return err;
+			}
+			topic[topic_sz] = '\0';
+			err = at_params_short_get(&at_param_list, 2, &op);
+			if (err < 0) {
+				return err;
+			}
+			err = at_params_short_get(&at_param_list, 3, &qos);
+			if (err < 0) {
+				return err;
+			}
+			err = do_mqtt_subscribe(op, topic, topic_sz, qos);
+		} else if (at_params_valid_count_get(&at_param_list) == 3) {
+			err = at_params_string_get(&at_param_list, 1,
+							topic, &topic_sz);
+			if (err < 0) {
+				return err;
+			}
+			topic[topic_sz] = '\0';
+			err = at_params_short_get(&at_param_list, 2, &op);
+			if (err < 0) {
+				return err;
+			}
+			err = do_mqtt_subscribe(op, topic, topic_sz, 0);
+		} else {
+			return -EINVAL;
+		}
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf, "#XMQTTSUB: <topic>, (0, 1), (0, 1, 2)\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+/**@brief API to handle MQTT AT commands
+ */
+int slm_at_mqtt_parse(const char *at_cmd)
+{
+	int ret = -ENOTSUP;
+	enum at_cmd_type type;
+
+	for (int i = 0; i < AT_MQTT_MAX; i++) {
+		if (slm_util_cmd_casecmp(at_cmd, m_mqtt_at_list[i].string)) {
+			ret = at_parser_params_from_str(at_cmd, NULL,
+						&at_param_list);
+			if (ret) {
+				LOG_ERR("Failed to parse AT command %d", ret);
+				return -EINVAL;
+			}
+			type = at_parser_cmd_type_get(at_cmd);
+			ret = m_mqtt_at_list[i].handler(type);
+			break;
+		}
+	}
+
+	return ret;
+}
+
+int slm_at_mqtt_init(void)
+{
+	return 0;
+}
+
+int slm_at_mqtt_uninit(void)
+{
+	client.broker = NULL;
+	fds.fd = INVALID_FDS;
+
```

Comment:
The thread needs to be terminated here

Reply:
As Robert said, the thread is blocked so no need to be terminated.